### PR TITLE
Internal: Tests - Playwright Widgets Regression Tests POC (WIP) [ED-7498]

### PR DESCRIPTION
**Pull request by ShayHel**

### CODE REVIEW

**Review by matipojo — CHANGES_REQUESTED** [review]
Please make Linux resolution same as win32

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: includes/base/controls-stack.php [hunk 1]
```diff
@@ -408,12 +408,14 @@ public function add_control( $id, array $args, $options = [] ) {
...
-			];
+		if ( $this->current_popover ) {
+			$args['popover'] = [];
+
+			if ( ! $this->current_popover['initialized'] ) {
+				$args['popover']['start'] = true;
 
-			$this->current_popover['initialized'] = true;
+				$this->current_popover['initialized'] = true;
+			}
```
Comment (StyleShit) [inline includes/base/controls-stack.php:418]: @matipojo 

It increases the widgets cache response from 2.84MB (241.62KB transferred) to 2.86MB (242.73KB transferred).

File: assets/dev/js/editor/controls/popover-toggle.js [hunk 2]
```diff
@@ -17,6 +17,13 @@ export default class ControlPopoverStarterView extends ControlChooseView {
 		} );
 	}
 
+	onShow() {
+		this.popover = this.$el.next( '.elementor-controls-popover' );
+
+		// Attach the current control as a toggle of its popover.
+		this.popover[ 0 ].dataset.popoverToggle = `elementor-control-default-${ this.model.cid }`;
+	}
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/popover-toggle.js:25]: @matipojo 

I use it in order to make it easier (and doable...) to select the popover and its toggle in the tests.

File: tests/playwright/utils/widgets/heading.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	async setup() {
+		// Set the parent section background to red in order to test controls such as "blend mode".
+		await this.editor.page.evaluate( () => {
+			const section = elementor.getContainer( 'document' ).children[ 0 ];
+
+			$e.run( 'document/elements/settings', {
+				container: section,
+				settings: {
+					background_background: 'classic',
+					background_color: '#F00',
```
Comment (matipojo) [inline tests/playwright/utils/widgets/heading.js:17]: Please choose a more readable color

File: tests/playwright/utils/widgets/heading.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			}
+
+			if ( [ 'blend_mode', 'typography_font_weight' ].includes( controlId ) ) {
+				await this.editor.page.waitForTimeout( 700 );
+			}
+
+			await assertionsCallback( controlId, currentControlValue );
+		} );
+
+		await control.teardown();
```
Comment (matipojo) [inline tests/playwright/utils/widgets/heading.js:40]: Call `super.testControl` with a callback for the `control.test`

File: tests/playwright/utils/widgets/text-editor.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+		// Wait for server render after switching between tabs.
+		if ( [ 'align' ].includes( controlId ) ) {
+			await this.editor.page.waitForTimeout( 1000 );
+		}
+
+		await control.test( async ( currentControlValue ) => {
+			await assertionsCallback( controlId, currentControlValue );
+		} );
+
+		await control.teardown();
```
Comment (matipojo) [inline tests/playwright/utils/widgets/text-editor.js:20]: Call super.testControl with a callback for the control.test

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/playwright/sanity/widgets.test.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+const {
+	Choose,
+	Select,
+	Textarea,
+} = require( '../utils/controls' );
+
+const { Registrar } = require( '../utils/registrar' );
+
+test.only( 'All widgets sanity test', async ( { page }, testInfo ) => {
```
Comment (matipojo) [inline tests/playwright/sanity/widgets.test.js:19]: `only` 😥

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/playwright/sanity/widgets.test.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+	const iframeWidgets = [
+		'audio',
+		'google_maps',
+		'video',
+	];
+
+	const emptyWidgets = [
+		'html',
+		'image-carousel',
+		'image-gallery',
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply tests/playwright/sanity/widgets.test.js]: huh?

File: tests/playwright/utils/widgets/heading.js (hunk 4 above)
Comment (StyleShit) [reply tests/playwright/utils/widgets/heading.js:40]: let's talk about it

File: tests/playwright/sanity/widgets.test.js (hunk 6 above)
Comment (StyleShit) [reply tests/playwright/sanity/widgets.test.js:19]: Forgot to push 🤦 
Good catch!

File: tests/playwright/utils/widgets/heading.js (hunk 3 above)
Comment (StyleShit) [reply tests/playwright/utils/widgets/heading.js]: Changed it to be only for blend mode

**Pull request merged**